rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,166 @@
+# [RFC] Add algorithm to dot_general's parameters in the StableHLO specification
+
+Status: Review<br/>
+Initial version: 03/15/2024<br/>
+Last updated: 03/15/2024<br/>
+Discussion thread: [GitHub](https://github.com/openxla/stablehlo/pull/2096)
+
+## Motivation
+
+We would like to allow explicitly selecting the algorithm used for individual
+`dot_general` instructions, enforcing a well-defined numeric precision. We think
+that the current precision values (`DEFAULT`, `HIGH`, `HIGHEST`) are not
+specific enough for this, because they mean different things between hardware
+platforms and even between GPU types.
+
+We also want to make the API flexible enough to support additional algorithms
+that might be faster on certain hardware
+([6xBF16](https://arxiv.org/pdf/1904.06376.pdf), 3xTF32, etc) that can't fit
+into the limited precision_config values of `DEFAULT`, `HIGH` and `HIGHEST`. For
+example "6xBF16" provides similar precision to F32, but about 2x performance on
+some GPUs.
+
+When an algorithm is not supported by an accelerator, the program should fail
+instead of implicitly falling back to another. It should be the responsibility
+of higher level frameworks such as JAX, to define the required algorithm for
+each accelerator that is used.
+
+## Proposed Specification change
+
+### dot_general
+
+#### Semantics
+
+*Add these lines:*
+
+`algorithm` defines the main properties of the algorithm used to implement the
+dot operation, which also defines the precision. If the algorithm value is set
+(`!= UNSET`), then the `precision_config` is ignored. If the algorithm value is
+not set (`= UNSET`), then the `precision_config` will take effect.
+
+In general the first two types in the algorithm name are `AType` and `BType` -
+the precisions that the LHS and RHS of the operation are rounded to, and the
+third one is `AccumType` - the accumulation type. These types are independent
```

Comment:
Does it mean that the output has the precision of `AccumType`?

Reply:
Not really, we could accumulate in f32, but save the result in f16 in the end.

It's like this:

```
AccumType accum = 0;
accum += (AType) a * (BType) b;
accum += (AType) a * (BType) b;
accum += (AType) a * (BType) b;
...
OutputStorageType output = accum;
```

So if the output storage type can represent every value of the AccumType, then it will have "the same precision",
But if not, then it will have less.

Reply:
Hmm I'm still somewhat confused. In the example you have, because `output = accum`, an implicit cast is made. But the actual output of the dot operation is still the `AccumType`.?

Or are you saying code would be like

OutputStorageType dot_result = `stable_hlo.dot...accum_type=AccumType` so the implicit conversion is there anyway?

Reply:
The input/output of the HLO level dot operation will be always the input/output storage types (=the types in the HLO).
But during the computation of the result another type may be used, which is denoted by AType/BType/AccumType.
For example if we could have a dot which has F32 for both inputs and the output, but internally the GPU will convert the inputs to TF32. (TF32_TF32_F32)
So similarly, if there are casts from the accumulator type to the output type, that should happen inside the "kernel" and not as a separate HLO operation or similar.

Reply:
ahh ack, thanks for the explanation. That makes sense.